File: ios/Sources/EdgeFrameworks/Core/EdgeKnowledgeCatalog.swift
public protocol EdgeKnowledgeCatalog: Sendable {
    func collections() async -> [EdgeKnowledgeCollection]

    func collection(
        id: String
    ) async -> EdgeKnowledgeCollection?

    func upsert(
        collection: EdgeKnowledgeCollection
    ) async throws

    func removeCollection(
        id: String
    ) async throws

    func sources(
        collectionID: String?
    ) async -> [EdgeKnowledgeSource]

    func source(
        id: String
    ) async -> EdgeKnowledgeSource?

    func upsert(
        source: EdgeKnowledgeSource
    ) async throws

    func removeSource(
        id: String
    ) async throws

    func removeAll() async throws
}

File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeKnowledgeCatalog.kt
package io.github.rajibsarwar.edgeframeworks

interface EdgeKnowledgeCatalog {
    suspend fun collections(): List<EdgeKnowledgeCollection>

    suspend fun collection(
        id: String
    ): EdgeKnowledgeCollection?

    suspend fun upsert(
        collection: EdgeKnowledgeCollection
    )

    suspend fun removeCollection(
        id: String
    )

    suspend fun sources(
        collectionId: String? = null
    ): List<EdgeKnowledgeSource>

    suspend fun source(
        id: String
    ): EdgeKnowledgeSource?

    suspend fun upsert(
        source: EdgeKnowledgeSource
    )

    suspend fun removeSource(
        id: String
    )

    suspend fun removeAll()
}
